### PR TITLE
Add group observers for Pixel and TargetedPixel

### PR DESCRIPTION
**Pull request by skuba31**
This PR adresses changes in group observers structure proposed in #325 reworked #330  to correctly include changes
Added class Observer0DGroup for working with groups of 0D observers
Changed names of methods and attributes including "sight_line" to "observer"
Added subclasses for Pixel and TargettedPixel observers
Changed spectroscopic groups to be subclasses of new Observer0DGroup
Added sight_line named wrappers for new observer attributes and methods to Spectroscopic groups for backward compatibility

### CODE REVIEW

**Comment by jacklovell** [comment]
Thanks for the contribution @skuba31. It'll take some time to review this properly and we're busy getting ready for a new release at the moment, so I'll take a proper look at this after the release is made.

**Comment by vsnever** [comment]
Hi @skuba31, thank you very much for this contribution and sorry for delayed response. 
I started reviewing the code and realised that it might be worth changing the concept a little.

After generalizing group observers to support basic 0D observers from Raysect, having special spectroscopic observers, namely `SpectroscopicFibreOptic` and `SpectroscopicSightLine`, in Cherab seems redundant to me. I think it would be simpler and more convenient if `SightLineGroup` and `FibreOpticGroup` supported the `SightLine` and `FibreOptic` observers from Raysect. Additional features of `SpectroscopicObserver0DGroup`, such as displaying the results of group observation and controlling the parameters of individual pipelines through group interfaces are applicable to all 0D observers and can be implemented in `Observer0DGroup`. To do this, the code from `_SpectroscopicObserver0DBase` must be moved to the corresponding functions/properties of `Observer0DGroup`. In this approach, we will have a base class `Observer0DGroup` with some additional functions and properties from `SpectroscopicObserver0DGroup` and 4 child classes: `PixelGroup`, `TargettedPixelGroup`, `SightLineGroup` and `FibreOpticGroup`, which use observers from Raysect.

The `origin` and `direction` properties of `_SpectroscopicObserver0DBase` do not simplify the user code in my opinion, because using `translate()` and `rotate_basis()` is just as easy, so I think these observers can be marked deprecated especially considering that specifying just the forward direction without the up direction will not work with polarised light as @Mateasek pointed out in another conversation.

@CnlPepper, @jacklovell, @mateasek, do you have any thoughts on this?

**Comment by Mateasek** [comment]
I'm definitely in favor of moving towards this. I never really liked the idea of shifting from the transforms to the origin and direction. Yes, transform is a tiny bit harder to get around in head and to get right, but it is the key building block of Raysect. Let's leave this to be done by external codes and packages.

Regarding the inheritance you proposed, it sound more "natural" to me. 

The only drawback is that the proposed changed are backwards incompatible, aren't they?

How would you like to continue? Do it in this PR, or make a new one?

**Comment by skuba31** [comment]
@vsnever thanks for the update. Option 3 is fine with me. Is there a deadline to be met?

**Comment by vsnever** [comment]
Thank you very much, @skuba31. There is no deadline.

File: cherab/tools/observers/group/pixel.py
Comment (vsnever) [inline cherab/tools/observers/group/pixel.py:47]: "The observers attribute of PixelGroup must be a list or ..."

File: cherab/tools/observers/group/sightline.py
Comment (vsnever) [inline cherab/tools/observers/group/sightline.py:25]: Similar to `FibreOpticGroup`, this class should inherit from `Observer0DGroup` and use `SightLine` fomr Raysect instead of `SpectroscopicSightLine` from Cherab.

File: cherab/tools/observers/group/base.py
Comment (vsnever) [inline cherab/tools/observers/group/base.py:412]: I missed this function in my review. I suggest it should be changed the following way:
```python
    def _get_same_pipelines(self, item):
        pipelines = []
        observers = []
        for observer in self._observers:
            if isinstance(item, int):
                if item < len(observer.pipelines):
                    pipelines.append(observer.pipelines[item])
                    observers.append(observer)
            elif isinstance(item, str):
                matching_pipelines = [pipeline for pipeline in observer.pipelines if pipeline.name == item]
                if len(matching_pipelines) == 1:
                    pipelines.append(matching_pipelines[0])
                    observers.append(observer)

        if len(pipelines) == 0:
            raise ValueError("Pipeline {} was not found for any observer in this {}.".format((item, self.__class__.__name__)))

        pipeline_types = set(type(pipeline) for pipeline in pipelines)
        if len(pipeline_types) > 1:
            raise ValueError("Pipelines {} have different types for different observers.".format(item))

        return pipelines, observers
```

File: cherab/tools/observers/group/base.py
Comment (skuba31) [reply cherab/tools/observers/group/base.py:47]: This could be problematic as the pipeline base class does not implement display_progress or accumulate. This was ok for the SpectroscopicGroup which was inteded for spectroscopic pipelines which have these implemented.

File: cherab/tools/observers/group/base.py
Comment (skuba31) [reply cherab/tools/observers/group/base.py:165]: The base class for Observer0D does not have such functionality, therefore, I think it is not a good idea to have this property in the group base class.

File: cherab/tools/observers/group/base.py
Comment (skuba31) [reply cherab/tools/observers/group/base.py:386]: The pipelines property should definitely be included. I think that the readability of the connect_pipelines could be improved. Should I implement it "as is" or should I try to improve it?

File: cherab/tools/observers/group/base.py
Comment (skuba31) [reply cherab/tools/observers/group/base.py:465]: I think this method is not well suited for the base class as this class can be used with other than spectroscopic pipelines. It seems to me that in this case it would make sense to have a standalone function taking Observer0DGroup or list of pipelines as a parameter.

File: cherab/tools/observers/group/base.py
Comment (skuba31) [reply cherab/tools/observers/group/base.py:414]: This method is definitely useful. I think it is worth consideration whether to include this visualisation/postprocessing method in a class which is inteded for control of observers or if it should be a standalone function in order to keep the base class smaller.

File: cherab/tools/observers/group/base.py
Comment (vsnever) [reply cherab/tools/observers/group/base.py:47]: Agreed. Let's remove this, it's not a problem to loop over the observer's pipelines and toggle these parameters manually if needed.

File: cherab/tools/observers/group/base.py
Comment (vsnever) [reply cherab/tools/observers/group/base.py:165]: Agreed.

File: cherab/tools/observers/group/base.py
Comment (vsnever) [reply cherab/tools/observers/group/base.py:386]: Yes, please improve this. Probably, `connect_pipelines()` should take a list of (`pipeline_class`, `pipeline_kwargs`) tuples. 

When doing the group observation, it will make sense having `display_progress` set to False in the default kwargs for the spectral pipelines.

File: cherab/tools/observers/group/base.py
Comment (vsnever) [reply cherab/tools/observers/group/base.py:412]: If plotting functions are moved outside the base class, this function must be public.

File: cherab/tools/observers/group/base.py
Comment (vsnever) [reply cherab/tools/observers/group/base.py:414]: Probably, moving this function and `plot_spectra()` outside the base class would do better. In this case they should be renamed to something like `plot_group_total()` and `plot_group_spectra()`.

File: cherab/tools/observers/group/base.py
Comment (skuba31) [reply cherab/tools/observers/group/base.py:386]: This seems as reasonable approach. But I am wondering if it would be possible to create interface so that group would be compatible with 1D pipelines, because the most common use case seems to be to have observers with the same pipelines.

Comment (vsnever) [reply cherab/tools/observers/group/base.py:386]: It is possible but you'll need to completely rewrite the `observe()` method. I'm not sure if this is a good approach. First, the current implementation allows different observers in the group have different spectral settings. If all 0D observers share the same 1D pipeline, this would not be possible. Second, I think this breaks the integrity. For example, you'll loose the observed frame, if you add a new observer to the group between observations even if the pipeline is in accumulation mode. 

However, if it is OK that all observers in a group have the same spectral properties and differ only in geometry, then the group observer can be implemented as `Observer1D`. I think it's easier than sharing 1D pipeline between multiple 0D observers. The disadvantage of this approach is that 1D observers do not have a `samples_per_task` property. Thus, if `MulticoreEngine` is used and the number of observers in the group is less than the number of parallel processes, rendering will not be efficiently parallelized.  Let's hear what @Mateasek, @jacklovell and @CnlPepper have to say about this.

Comment (vsnever) [reply cherab/tools/observers/group/base.py:386]: Actually, choosing between keeping as is or converting to `Observer1D`, I prefer to leave it as is.

Comment (skuba31) [reply cherab/tools/observers/group/base.py:386]: I see, let's keep it as it is now.

**Comment by vsnever** [comment]
@skuba31, I would prefer to keep the `SpectroscopicFibreOptic` and `SpectroscopicSightLine` in the code for a couple of releases. Otherwise we break the user space without warning. Just add something like:
```python
"""
.. deprecated:: 1.4.0
       Use Raysect's FibreOptic observer instead.
""" 
```
in the beginning of their docstrings.

File: cherab/tools/observers/group/base.py
Comment (skuba31) [reply cherab/tools/observers/group/base.py:412]: I have moved this function into plotting module.

File: cherab/tools/observers/group/targettedpixel.py
Comment (vsnever) [inline cherab/tools/observers/group/targettedpixel.py:80]: Correct `len(self._sight_lines)` to `len(self._observers)`.

File: cherab/tools/observers/group/targettedpixel.py
Comment (vsnever) [inline cherab/tools/observers/group/targettedpixel.py:97]: As above.

File: cherab/tools/observers/group/targettedpixel.py
Comment (vsnever) [inline cherab/tools/observers/group/targettedpixel.py:134]: As above.

File: cherab/tools/observers/group/targettedpixel.py
Comment (vsnever) [inline cherab/tools/observers/group/targettedpixel.py:120]: The correct property name is `pixel.targets`, not `pixel.target`. `TargettedPixel` receives a list of targets, not a single target, so here we have to check if the supplied list contains lists/tuples or primitives. The target primitives must be connected to the same scene graph as the `TargettedPixel`, which means they must have the same root, but changing the parent of the target will affect coordinate space transformations, so the parent should not be changed. I propose the following implementation: 
```python
    @targets.setter
    def targets(self, value):
        if all(isinstance(v, (list, tuple)) for v in value):
            if len(value) == len(self._observers):
                for pixel, v in zip(self._observers, value):
                    pixel.targets = v
            else:
                raise ValueError("The length of 'value' ({}) "
                                 "mismatches the number of pixels ({}).".format(len(value), len(self._observers)))
        else:
            # assuming a list of primitives, the pixel's setter will throw an error if not
            for pixel in self._observers:
                pixel.targets = value
```

**Review by vsnever — CHANGES_REQUESTED** [review]
Thank you @skuba31, I really like your implementation. I have one major comment on `TargettedPixelGroup` and a few minor correction to some docstrings and error messages.

Could you also add `TargettedPixelGroup` and `PixelGroup` to docs/source/tools/observers.rst?

File: cherab/tools/observers/group/base.py
Comment (skuba31) [reply cherab/tools/observers/group/base.py:29]: Moved the deprecated indication to the right base.py (with definition of _SpectroscopicObserver0DBase)

File: cherab/tools/observers/group/spectroscopic.py
Comment (skuba31) [reply cherab/tools/observers/group/spectroscopic.py:30]: I kept the class so that we can have SpectroscopicGroup for backward compatibility. Sublcasses were also added to this file so that all the depracated classes would be here.

**Comment by skuba31** [comment]
Thanks @vsnever for the review. I have modified the code and the docs.

**Review by vsnever — APPROVED** [review]
Thanks @skuba31, now I'm happy with this PR.

File: cherab/tools/observers/group/plotting.py
Comment (jacklovell) [inline cherab/tools/observers/group/plotting.py:37]: Raise an `IndexError` here: you can then explain exactly why no pipeline was found. You can either use this check or rely on the `IndexError` which would be returned when trying to access `observer.pipelines[item]`. Also, what happens if the index is negative? Is it what the user would expect? Especially if `abs(index) > len(observer.pipelines)`: you'll get an `IndexError` anyway.

File: cherab/tools/observers/group/plotting.py
Comment (jacklovell) [inline cherab/tools/observers/group/plotting.py:42]: Ff there is more than 1 matching pipeline, this check will silently fail. Best to invert the logic in the check: if there is not exactly 1 matching pipeline then raise an error.

File: cherab/tools/observers/group/plotting.py
Comment (jacklovell) [inline cherab/tools/observers/group/plotting.py:83]: Use `group.observers`

**Review by jacklovell — CHANGES_REQUESTED** [review]
Thanks for this @skuba31. It's looking pretty good from my end: I agree with what @vsnever has already suggested and have only a few suggestions myself.

Could you add some tests for the new observers? It can be simple things like checking assignment and retrieval of properties, and that invalid inputs are correctly detected. Aim for as much coverage of the new classes as possible, within reason.

I notice there's quite a lot of repetition in each of the sub-classes, in particular in the `observers` setter and the `add_observer` method. The only thing that differs for each sub-class is the type of the observer to add: this can be parametrised by a class variable instead. Something like the following:
```python
class TargettedPixelGroup(Observer0DGroup):
    _OBSERVER_TYPE = TargettedPixel
    ...

class SightlineGroup(Observer0DGroup):
    _OBSERVER_TYPE = Sightline
    ...
```
Then the relevant part of the setter in `Observer0DGroup` becomes:
```python
@observers.setter
def observers(self, value):
    ...  # Check type of value as before
    for observer in value:
        if not isinstance(observer, self._OBSERVER_TYPE):
            raise TypeError("Observer should be of type {}".format(self._OBSERVER_TYPE))
    # etc.
```
You can then do the same with the `add_observer` method. Define both of these only in the base class, then if we need to add new functionality, new checks or bug fixes they only need to be done in one place.

**Comment by jacklovell** [comment]
Oh, almost forgot. Can you add an entry to the changelog too for these new features? It'll be for version 1.4.0 (release date TBD).

File: cherab/tools/observers/group/plotting.py
Comment (vsnever) [reply cherab/tools/observers/group/plotting.py:37]: I agree with the comment on negative indexing, but I don't think that the `IndexError` should be raised. This function returns the lists of matching pipelines and the observers for which those pipelines are found. The observers in a group can have different number of pipelines, and if the _i_-th pipeline cannot be found for the _k_-th observer because this _k_-th observer has _N<sub>k</sub>_ < _i_ number of pipelines, then the _k_-th observer should be skipped because it has no matching pipeline.

File: cherab/tools/observers/group/plotting.py
Comment (vsnever) [reply cherab/tools/observers/group/plotting.py:42]: Partially agree, the error should be raised but only if `len(matching_pipelines) > 1`, because in this case we don't know which pipeline to choose. If there is no matching pipeline for this observer, it should be skipped.

File: cherab/tools/observers/group/plotting.py
Comment (vsnever) [inline cherab/tools/observers/group/plotting.py:137]: Sorry, I missed this. I think similarly to `plot_group_total()` we should check if the name of the observer is set, and if not, use its index as a label.
```python
label = observer.name if observer.name and len(observer.name) else group.observers.index(observer)
```

File: cherab/tools/observers/group/plotting.py
Comment (skuba31) [reply cherab/tools/observers/group/plotting.py:37]: I have added a warning if invalid index is used.

File: cherab/tools/observers/group/plotting.py
Comment (skuba31) [reply cherab/tools/observers/group/plotting.py:42]: I have added a warning if no matching pipeline was found. An error is raised only if there are more pipelines.

**Comment by skuba31** [comment]
@jacklovell I have changed the observers setter and add_observer methods. The tests might take a bit longer as this is something I have little experience with and I might not have enougt time in the next few weeks.  
Regarding the changelog, I will have to synchronize with the new development branch as this PR started with older version. Should I reset the history and create a new PR after that, or shoul I keep this one?

**Comment by jacklovell** [comment]
It's fine to keep working in this branch. Just merge the changes from development into your branch. Something like this, assuming your `enhancement/TargetedPixel` branch is the current working branch:
```bash
git fetch origin development:development
git merge development
```
Best to do this before adding your changelog entry, otherwise you'll have merge conflicts to resolve. 

If you're not confident doing this then just add the changelog anyway, and I'll handle the conflicts and re-sync with the development branch when the pull request is merged. It shouldn't be hard to do, as development and your branch haven't diverged significantly.

Regarding tests, have a look at some of the examples in cherab/tools/tests: they should give you an idea. I don't think we specifically have tests for other observers yet (I should probably add some for the bolometry...), but something like `test_plasma_slabs` is a good example.

**Comment by skuba31** [comment]
@jacklovell I have merged the development branch using squash commit and added the unit tests for groups.

**Comment by jacklovell** [comment]
It would have been better just to do a normal merge (not squashed) of development, as the squash commit seems to have sent Github mad and added a huge number of changes to the PR. It's not a big problem as the change set is sane when I check it out locally, and it'll clean up once it's merged. Unit tests look promising: I'll take a proper look at this next week but assuming no major surprises it should be good to merge.

**Comment by jacklovell** [comment]
@skuba31 the tests look good: coverage is excellent and it seems they highlighted a few small bugs which are now fixed. I see there's no test coverage for the plotting though. This is fair enough as it's difficult to test plotting with automated unit tests. I think a demo would be good to add, both to illustrate how to do the plotting and also as a quick check that the functions work. You can put an example in `demos/observers`. Have a look at the existing demos for inspiration.

**Comment by vsnever** [comment]
Hi @skuba31, I think the easiest way to create the demo requested by @jacklovell is to use the [generomak_raw_files_solps_line_emitter.py](https://github.com/cherab/solps/blob/master/demos/generomak_raw_files_solps_line_emitter.py) demo from Cherab-SOLPS. Just replace:
```python
from cherab.solps import load_solps_from_raw_output
...
demos_directory = os.path.dirname(__file__)
simulation_directory = os.path.join(demos_directory, 'data', 'raw')
print('Loading simulation...')
sim = load_solps_from_raw_output(simulation_directory)
print('Creating plasma...')
plasma = sim.create_plasma()
```
with
```python
from cherab.generomak.plasma import get_edge_plasma
...
plasma = get_edge_plasma()
```
remove this caching:
```python
# Caching the H-alpha radiation power density on the SOLPS mesh
min_wavelength = 654.
max_wavelength = 658.
print('Caching H-alpha radiation power density on SOLPS mesh...')
direction = Vector3D(0, 0, 1)
halpha_radiation = np.zeros_like(sim.electron_density)  # total radiation density on SOLPS mesh
spectrum = Spectrum(min_wavelength, max_wavelength, 1)
for ix in range(halpha_radiation.shape[1]):
    for iy in range(halpha_radiation.shape[0]):
        point = Point3D(sim.mesh.cr[iy, ix], 0, sim.mesh.cz[iy, ix])
        for model in plasma.models:
            halpha_radiation[iy, ix] += 4. * np.pi * model.emission(point, direction, spectrum.new_spectrum()).total()

sim.halpha_total_radiation = halpha_radiation

# Replace plasma models with the SOLPSLineEmitter
plasma.models = [SOLPSLineEmitter(sim.halpha_total_radiation_f2d, h_alpha)]
```
and this plotting:
```python
# Plotting the projection of sight lines on the poloidal plane.
plot_emission_and_sightlines(sim, group)
```
Then update the respective code for the deprecated `FibreOpticGroup` with the code for the new `FibreOpticGroup`. 

I'll update the `generomak_raw_files_solps_line_emitter.py` to use the new `FibreOpticGroup` once this PR is merged.

**Comment by skuba31** [comment]
I have added the demo for group observers. But I had to make changes in [cherab.generomak.plasma.plasma.py](https://github.com/cherab/core/blob/0eb7454137b17bfbeca6d5e3eae4281f9cd4c157/cherab/generomak/plasma/plasma.py)
[line 63](https://github.com/cherab/core/blob/0eb7454137b17bfbeca6d5e3eae4281f9cd4c157/cherab/generomak/plasma/plasma.py#L63)
```python
profiles_dir = os.path.join(os.path.dirname(__file__), "/data/plasma/edge")
```
to
```python
profiles_dir = os.path.join(os.path.dirname(__file__), "data/edge")
```

[line 178](https://github.com/cherab/core/blob/0eb7454137b17bfbeca6d5e3eae4281f9cd4c157/cherab/generomak/plasma/plasma.py#L178)
```python
profiles_dir = os.path.join(os.path.dirname(__file__), "data/plasma/edge")
```
to
```python
profiles_dir = os.path.join(os.path.dirname(__file__), "data/edge")
```

Should I include these changes in this PR as well?

**Comment by vsnever** [comment]
> Should I include these changes in this PR as well?

Yes please do it. We argued about where to put the data in https://github.com/cherab/core/pull/337, so it's not surprising that we ended up missing this. Thanks for the demo, it's the first demo to use the Generomak edge plasma.

**Comment by skuba31** [comment]
@vsnever It should be fixed now.

**Comment by vsnever** [comment]
@jacklovell, don't merge this yet. The results of the demo do not match those of the original cherab-solps demo. The plasma profiles should be the same, so I'm looking into what's causing this.

**Comment by vsnever** [comment]
> @jacklovell, don't merge this yet. The results of the demo do not match those of the original cherab-solps demo. The plasma profiles should be the same, so I'm looking into what's causing this.

Everything works fine with the SOLPS plasma, so something is wrong with `get_edge_plasma()` or the json data.

**Comment by vsnever** [comment]
Found it!
@skuba31, I'm sorry you have to fix our bugs, but could you please fix a copy-paste error in `get_edge_interpolators()`:
https://github.com/cherab/core/blob/0eb7454137b17bfbeca6d5e3eae4281f9cd4c157/cherab/generomak/plasma/plasma.py#L104
There should be "density" of course.
```python
    ne = Discrete2DMesh.instance(te, profiles["electron"]["density"], limit=False)
```
Also, I don't understand why the geometry transform is done like this
https://github.com/cherab/core/blob/0eb7454137b17bfbeca6d5e3eae4281f9cd4c157/cherab/generomak/plasma/plasma.py#L195
I think it should be
```python
    geometry_transform = translate(0, 0, z_range[0])
```
@Mateasek can you look at this?

**Comment by skuba31** [comment]
@vsnever I have implemented the suggested changes. The results of the demo have changed, but I was not able to verify them, by running the original solps demo.

**Comment by vsnever** [comment]
Yes, I'm happy with this PR. Great job @skuba31!

**Comment by jacklovell** [comment]
Merged. Thanks for the great contribution @skuba31!

**Comment by skuba31** [comment]
@jacklovell @vsnever Thank you for the valuable feedback.

**Pull request merged**
